### PR TITLE
Move LobbyWatcherThread ownership to ServerModel, prep for ServerPoller

### PR DESCRIPTION
**Pull request by DanVanAtta**
Change summary:
- Move LobbyWatcherThread ownership from SetupPanel to ServerModel
- Remove unused moderator socket remote actions
- Extract headless 'LaunchAction' anonymous inner class to its own class

ServerPoller will be a background thread on hosts to check for recently
banned players. This will allow moderators to kick/ban players from hosted
games.

To get there we need to have access to both ServerMessengers and the API
key used in LobbyWatcher-Thread. The place where we create lobby-watcher-thread
will be the same where we will create the upcoming server-poller.

 

## Functional Changes

[] New map or map update
[] New Feature
[] Feature update or enhancement
[] Feature Removal
[] Code Cleanup or refactor
[] Configuration Change
[] Problem fix:  
[] Other:   

## Testing

[x] Manual testing done
- launched a headless server, connected to it and started a game, set password, and for eachverified lobby status was updated
- hosted a game on local lobby with game-headed, verify game status was updated in lobby as maps changed or game was started

### CODE REVIEW

File: game-core/src/main/java/org/triplea/game/server/HeadlessLaunchAction.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  @Override
+  public IDisplay startGame(
+      final LocalPlayers localPlayers,
+      final IGame game,
+      final Set<Player> players,
+      final Chat chat) {
+    final UiContext uiContext = new HeadlessUiContext();
+    uiContext.setDefaultMapDir(game.getData());
+    uiContext.setLocalPlayers(localPlayers);
+    return new HeadlessDisplay();
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-core/src/main/java/org/triplea/game/server/HeadlessLaunchAction.java]: (Just realized this is just moved code and I have the suspicion that the original code was written by myself, but this should be fixed nevertheless, UiContext seems to be unused here.)

File: game-core/src/main/java/games/strategy/engine/framework/startup/LobbyWatcherThread.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  @Getter private final InGameLobbyWatcherWrapper lobbyWatcher = new InGameLobbyWatcherWrapper();
+  @Nonnull private final GameSelectorModel gameSelectorModel;
+  @Nonnull private final IServerMessenger serverMessenger;
+  @Nonnull private final WatcherThreadMessaging watcherThreadMessaging;
+
+  public void createLobbyWatcher(
+      final URI lobbyUri, final GameHostingResponse gameHostingResponse) {
+
+    final HttpLobbyClient lobbyClient =
+        HttpLobbyClient.newClient(lobbyUri, ApiKey.of(gameHostingResponse.getApiKey()));
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/engine/framework/startup/LobbyWatcherThread.java]: The getters are significant for JSON decode/encode.  Changing to return an object type and we change the JSON, the tests we have begin to fail with:

```
feign.codec.DecodeException: java.lang.IllegalStateException: Expected a string but was BEGIN_OBJECT at line 1 column 42 path $.apiKey
```

I try to subscribe to a principle where inter-system or inter-service dependencies should be low-level types, like 'string', 'int', and avoid objects. 'intra-system' dependencies on the other hand should be strongly typed. IE: when receiving data from external sources, do not depend on object types, once received, convert to an object type at the edge of the system and then use strong types within the system.

Hence, it is by design that "all usages of the getter convert the string into an ApiKey object afterwards"

File: game-core/src/main/java/games/strategy/engine/framework/startup/LobbyWatcherThread.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+              lobbyWatcher.setInGameLobbyWatcher(watcher);
+
+              LocalServerAvailabilityCheck.builder()
+                  .connectivityCheckClient(lobbyClient.getConnectivityCheckClient())
+                  .localPort(serverMessenger.getLocalNode().getPort())
+                  .errorHandler(watcherThreadMessaging::serverNotAvailableHandler)
+                  .build()
+                  .run();
+
+              System.clearProperty(TRIPLEA_NAME);
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/engine/framework/startup/LobbyWatcherThread.java]: Extreme tecnnical debt, this is moved code. If this property is set when triplea starts, it thinks we're loading a save-game IIRC. Hence, we when we fire up a new process that we do not pass through that logic gate and go to a game hosting.

File: game-core/src/main/java/games/strategy/engine/framework/startup/WatcherThreadMessaging.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                        + "\nDo you want to view the tutorial on how to host? "
+                        + "This will open in your internet browser.",
+                    "View Help Website?",
+                    JOptionPane.YES_NO_OPTION)
+                == JOptionPane.YES_OPTION) {
+              OpenFileUtility.openUrl(UrlConstants.USER_GUIDE);
+            }
+          });
+    }
+  }
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/engine/framework/startup/WatcherThreadMessaging.java]: I was tempted to do that. The reason for same-file is largely due to excessive classes in this package, I was concerned the relationship between the classes would get lost as they would not sort together. We'd wind up with a situation where some classes in the package are related, others are not, and others are related in but another way. I see a cohesion problem there, the classes being split would make that worse. I figure it's easy enough if we ever can get more items in `game-headless` that we can split then. I've been thinking we may *need* dependency injection to really be able to make progress on that kind of slit. On another level, it's becoming a bit more convincing that a lot of code just needs to be re-done.

File: game-core/src/main/java/games/strategy/engine/framework/startup/mc/ServerModel.java [hunk 5]
```diff
@@ -383,23 +388,38 @@ public void createServerMessenger() {
...
           new ServerMessenger(props.getName(), props.getPort(), objectStreamFactory);
-      final ClientLoginValidator clientLoginValidator = new ClientLoginValidator(serverMessenger);
-      clientLoginValidator.setGamePassword(props.getPassword());
-      serverMessenger.setLoginValidator(clientLoginValidator);
       serverMessenger.addConnectionChangeListener(this);
 
       messengers = new Messengers(serverMessenger);
       messengers.registerRemote(serverStartupRemote, SERVER_REMOTE_NAME);
+
+      GameHostingResponse gameHostingResponse = null;
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/engine/framework/startup/mc/ServerModel.java]: Fair enough, done: 30d089b12

File: game-core/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java [hunk 6]
```diff
@@ -551,13 +464,7 @@ public ChatModel getChatModel() {
...
   @Override
   public synchronized Optional<ILauncher> getLauncher() {
-    return model
-        .getLauncher()
-        .map(
-            launcher -> {
-              launcher.setInGameLobbyWatcher(lobbyWatcher);
-              return launcher;
-            });
+    return model.getLauncher().map(launcher -> launcher);
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java]: I was a bit surprised, but there was, the types do not line up without it. You should see a compile error without it.

File: game-core/src/main/java/org/triplea/game/server/HeadlessLaunchAction.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/org/triplea/game/server/HeadlessLaunchAction.java]: Are there any side-effects in `HeadlessUiContext` that we need?

Indeed, this is moved code. I'm not willing to complicate this change/branch further, it was a small miracle it was landed as is and did not involve more changes.

[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/org/triplea/game/server/HeadlessLaunchAction.java]: > it was a small miracle 

The circular dependencies and interactions between the launchers, models, panels, are very intertwined. Fixing it properly is a challenge, I'm tempted to be convinced that it needs a re-write to be fixed.

File: game-core/src/main/java/games/strategy/engine/framework/startup/mc/ServerModel.java [hunk 7]
```diff
@@ -383,23 +388,38 @@ public void createServerMessenger() {
...
+                    : new WatcherThreadMessaging.HeadedWatcherThreadMessaging(ui));
+        lobbyWatcherThread.createLobbyWatcher(lobbyUri, gameHostingResponse);
+      }
+
       chatController = new ChatController(CHAT_NAME, messengers, node -> false);
 
       if (ui == null) {
         chatModel =
             new HeadlessChat(new Chat(new MessengersChatTransmitter(CHAT_NAME, messengers)));
-        chatModelCancel = Runnables.doNothing();
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/engine/framework/startup/mc/ServerModel.java]: The 'createServerMessenger' seems to have been a hack to deal with circular dependencies. I'm not sure how much we should 'honor' the hack, (we'd run into other problems if this method were called twice), but you make a fair point:
9e621a855968df60fe7ee88ac51bc58c88506677

Looking at it more, looks like we can properly undo the hack now and combine the 'create' method with the constructor: 5fa687ab268cbebfc655c6664c87a233d0713708

File: game-core/src/main/java/org/triplea/game/server/HeadlessServerSetup.java [hunk 8]
```diff
@@ -134,13 +58,7 @@ ServerModel getModel() {
...
   @Override
   public synchronized Optional<ILauncher> getLauncher() {
-    return model
-        .getLauncher()
-        .map(
-            launcher -> {
-              launcher.setInGameLobbyWatcher(lobbyWatcher);
-              return launcher;
-            });
+    return model.getLauncher().map(launcher -> launcher);
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/org/triplea/game/server/HeadlessServerSetup.java]: Results in a compile error, expected type `Optional<ILauncher>` , actual, `Optional<IServerLauncher>`

File: game-core/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java]: If we change the returned interface type, then we can do this: 5d5337a8af2c1ac3cd7b8ea74551ad96e3813fa7

File: game-core/src/main/java/org/triplea/game/server/HeadlessLaunchAction.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-core/src/main/java/org/triplea/game/server/HeadlessLaunchAction.java]: > Are there any side-effects in `HeadlessUiContext` that we need?

I wouldn't think so but I could be wrong. AFAIK `UiContext` does just hold some information for later use. Also had a quick look at the implementation, only reassuring my thoughts.

> it was a small miracle it was landed as is and did not involve more changes.

Well I did put at least some effort into separating those classes a little bit more in #5070 maybe this made it simpler to deal with.

File: game-core/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java]: Thanks for changing the interface type 👍

File: game-core/src/main/java/games/strategy/engine/framework/startup/WatcherThreadMessaging.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/engine/framework/startup/WatcherThreadMessaging.java]: Good Point.

**Comment by DanVanAtta** [comment]
The method it does not like was 'touched', so it is complaining. Code climate checks do not need to pass for merge. 

I'm thinking the rule is probably going to be:
- new code should pass code climate
- existing code with issues should be a best effort to fix
- if existing code is flagging the problem and was barely modified, it's okay leaving it alone.

In this case refactoring `createServerMessenger` introduces risk, I'd hate to introduce a bug at this late of a stage.

**Pull request merged**